torch/lib/c10d/ProcessGroupMPI.hpp

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include "ProcessGroup.hpp"
+#include "Types.hpp"
+#include "Utils.hpp"
+
+#include <mpi.h>
+
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+namespace c10d {
+
+// WorkEntry is the state associated with a single MPI run instance.
+// It include the source Tensor list and destination Tensor list, as well as
+// The actual run function that will operate either on src or dst or both.
+struct WorkEntry {
+  explicit WorkEntry(
+      std::vector<at::Tensor>* src,
+      std::vector<at::Tensor>* dst,
+      std::function<void(std::unique_ptr<WorkEntry>&)> run)
+      : src(src), dst(dst), run(run) {}
+
+  // Not copyable
+  WorkEntry(const WorkEntry&) = delete;
+  // Not copy assignable
+  WorkEntry& operator=(const WorkEntry&) = delete;
+
+  // For input and output tensors (in-place), we will always use src
+  std::vector<at::Tensor>* src;
+  std::vector<at::Tensor>* dst;
+  std::function<void(std::unique_ptr<WorkEntry>&)> run;
+};
```

Comment:
This has the potential to conflict with other symbols in c10d in the future. Let's address this in a separate PR after discussing this: we can either namespace the backends, or scope these backend specific classes with the specific process group class (e.g. ProcessGroupMPI::WorkEntry). I prefer the former because it is more terse.

@teng-li @apaszke Thoughts?

Reply:
Works for me. Will also move the Gloo AlgorithmKey and AlgorithmValue structs inside the class def.